olm/catalogd/v1alpha1/clustercatalog_types.go

```diff
@@ -0,0 +1,219 @@
+	//   - Status: a string representation of the state of the condition. Can be one of ["True", "False", "Unknown"].
+	//   - Reason: a string representation of the reason for the current state of the condition. Typically useful for building automation around particular Type+Reason combinations.
+	//   - Message: a human-readable message that further elaborates on the state of the condition.
+	//
+	// The current set of condition types are:
+	//   - "Serving", which represents whether or not the contents of the catalog are being served via the HTTP(S) web server.
+	//   - "Progressing", which represents whether or not the ClusterCatalog is progressing towards a new state.
+	//
+	// The current set of reasons are:
+	//   - "Succeeded", this reason is set on the "Progressing" condition when progressing to a new state is successful.
+	//   - "Blocked", this reason is set on the "Progressing" condition when the ClusterCatalog controller has encountered an error that requires manual intervention for recovery.
+	//   - "Retrying", this reason is set on the "Progressing" condition when the ClusterCatalog controller has encountered an error that might be resolvable on subsequent reconciliation attempts.
+	//   - "Available", this reason is set on the "Serving" condition when the contents of the ClusterCatalog are being served via an endpoint on the HTTP(S) web server.
+	//   - "Unavailable", this reason is set on the "Serving" condition when there is not an endpoint on the HTTP(S) web server that is serving the contents of the ClusterCatalog.
+	//
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+	// resolvedSource contains information about the resolved source based on the source type.
+	//
+	// Below is an example of a resolved source for an image source:
+	// resolvedSource:
+	//
+	//  image:
+	//    lastSuccessfulPollAttempt: "2024-09-10T12:22:13Z"
+	//    ref: quay.io/operatorhubio/catalog@sha256:c7392b4be033da629f9d665fec30f6901de51ce3adebeff0af579f311ee5cf1b
+	//  type: Image
+	// +optional
+	ResolvedSource *ResolvedCatalogSource `json:"resolvedSource,omitempty"`
+	// urls contains the URLs that can be used to access the catalog.
+	// +optional
+	URLs *CatalogURLs `json:"urls,omitempty"`
+	// lastUnpacked represents the time when the
+	// ClusterCatalog object was last unpacked successfully.
+	// +optional
+	LastUnpacked metav1.Time `json:"lastUnpacked,omitempty"`
+}
+
+type CatalogURLs struct {
+	// base is a required cluster-internal URL from which on-cluster components can access the API endpoint for this catalog
+	Base string `json:"base"`
+}
+
+// CatalogSource is a discriminated union of possible sources for a Catalog.
+// CatalogSource contains the sourcing information for a Catalog
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type == 'Image' && has(self.image)",message="source type 'Image' requires image field"
```

Comment:
This forces the value of type to required, which is fine, but, would need to be changed if you ever added a new value

We typically use
```suggestion
// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Image' ?has(self.image) : !has(self.image)",message="image is required when type is Image, and forbidden otherwise"
```
You can then copy/paste this for new fields in the union without changing this one

Reply:
Nice! Thanks for this suggestion. That will be helpful for having a reasonable "message" when we add new types.